MapboxDirections/OfflineDirections.swift

```diff
@@ -0,0 +1,139 @@
+     */
+    @discardableResult
+    func downloadTiles(for boundingBox: BoundingBox, version: Version, progressHandler: @escaping OfflineDownloaderProgressHandler, completionHandler: @escaping OfflineDownloaderCompletionHandler) -> URLSessionDownloadTask
+}
+
+extension Directions: OfflineDirectionsProtocol, URLSessionDownloadDelegate {
+    
+    func availableVersionsURL() -> URL {
+        
+        let url = apiEndpoint.appendingPathComponent("route-tiles/v1").appendingPathComponent("versions")
+        var components = URLComponents(string: url.absoluteString)
+        components?.queryItems = [URLQueryItem(name: "access_token", value: accessToken)]
+        
+        return components!.url!
+    }
+    
+    func tilesURL(for boundingBox: BoundingBox, version: Version) -> URL {
+        
+        let url = apiEndpoint.appendingPathComponent("route-tiles/v1").appendingPathComponent(boundingBox.path)
+        var components = URLComponents(string: url.absoluteString)
+        components?.queryItems = [URLQueryItem(name: "version", value: version.versionString),
+                                  URLQueryItem(name: "access_token", value: accessToken)]
+        
+        return components!.url!
+    }
+
+    @discardableResult
+    @objc
+    public func availableOfflineVersions(completionHandler: @escaping OfflineVersionsHandler) -> URLSessionDataTask {
+        
+        return URLSession.shared.dataTask(with: availableVersionsURL()) { (data, response, error) in
+            if let error = error {
+                return completionHandler(nil, error)
+            }
+            
+            guard let data = data else {
+                return completionHandler(nil, error)
+            }
+            
+            do {
+                let versionResponse = try JSONDecoder().decode(AvailableVersionsResponse.self, from: data)
+                completionHandler(versionResponse.availableVersions, error)
+            } catch {
+                completionHandler(nil, error)
+            }
+        }
+    }
+    
+    @discardableResult
+    @objc
+    public func downloadTiles(for boundingBox: BoundingBox, version: Version, progressHandler: @escaping OfflineDownloaderProgressHandler, completionHandler: @escaping OfflineDownloaderCompletionHandler) -> URLSessionDownloadTask {
+        
+        self.offlineProgressHandler = progressHandler
+        self.offlineCompletionHandler = completionHandler
+        
+        let configuration = URLSessionConfiguration.default
+        let url = tilesURL(for: boundingBox, version: version)
+        let session = URLSession(configuration: configuration, delegate: self, delegateQueue: nil)
+        
+        return session.downloadTask(with: url)
+    }
+    
+    public func urlSession(_ session: URLSession, downloadTask: URLSessionDownloadTask, didResumeAtOffset fileOffset: Int64, expectedTotalBytes: Int64) {
+        // TODO: resume download when range headers are supported
```

Comment:
Can we add support for this in advance of the feature availability on the server?